src/java.base/unix/native/libnet/SocketImpl.c

```diff
@@ -58,3 +59,114 @@ Java_java_net_SocketCleanable_cleanupClose0(JNIEnv *env, jclass c1, jint fd)
     NET_SocketClose(fd);
 }
 
+static jclass isa_class = NULL;
+static jmethodID isa_ctor = NULL;
+
+JNIEXPORT void JNICALL
+Java_java_net_Socket_initNative(JNIEnv *env, jclass c1)
+{
+    jclass isa_class_local = (*env)->FindClass(env, "java/net/InetSocketAddress");
+    if (isa_class_local == NULL) {
+        JNU_ThrowClassNotFoundException(env, "java.net.InetSocketAddress");
+        return;
+    }
+    isa_class = (*env)->NewGlobalRef(env, isa_class_local);
+    isa_ctor = (*env)->GetMethodID(env, isa_class, "<init>", "(Ljava/net/InetAddress;I)V");
+    if (isa_ctor == NULL) {
+        JNU_ThrowByName(env, "java/lang/NoSuchMethodError", "InetSocketAddress.<init>(java.net.InetAddress, int)");
+    }
+}
+
+static jobject create_isa(JNIEnv *env, jclass isa_class, jmethodID isa_ctor, SOCKETADDRESS *addr) {
+    jint port;
+    jobject inetAddr = NET_SockaddrToInetAddress(env, addr, &port);
+    return (*env)->NewObject(env, isa_class, isa_ctor, inetAddr, port);
+}
+
+JNIEXPORT jobjectArray JNICALL
+Java_java_net_Socket_getAddresses(JNIEnv *env, jclass cl, jint fd) {
+    int family;
+    socklen_t famlen = sizeof(int);
+    if (getsockopt(fd, SOL_SOCKET, SO_DOMAIN, &family, &famlen) != 0) {
+        JNU_ThrowByName(env, "java/net/SocketException", "Cannot find socket family");
+        return NULL;
+    } else if (family != AF_INET && family != AF_INET6) {
+        return NULL;
+    }
+
+    jobjectArray arr = (*env)->NewObjectArray(env, 2, isa_class, NULL);
+    if (arr == NULL) {
+        JNU_ThrowOutOfMemoryError(env, "java.net.InetSocketAddres[2]");
+        return NULL;
+    }
+
+    jobject localAddr;
+    SOCKETADDRESS local;
+    socklen_t llen = sizeof(SOCKETADDRESS);
+    if (getsockname(fd, &local.sa, &llen) != 0) {
+        JNU_ThrowIllegalArgumentException(env, strerror(errno));
+        return NULL;
+    } else {
+        localAddr = create_isa(env, isa_class, isa_ctor, &local);
+        if (localAddr == NULL) {
+           JNU_ThrowOutOfMemoryError(env, "java.net.InetSocketAddres");
+           return NULL;
+        }
+    }
```

Comment:
No need for `} else {` here and everywhere else since the previous block has anyway terminated the function.

This will make the code more streamlined.

```suggestion
}

localAddr = create_isa(env, isa_class, isa_ctor, &local);
if (localAddr == NULL) {
    JNU_ThrowOutOfMemoryError(env, "java.net.InetSocketAddres");
    return NULL;
}
```

Reply:
The code was supposed to mirror closer the remote part where we can't do that, but OK.